cpp/ql/lib/semmle/code/cpp/models/interfaces/FunctionInputsAndOutputs.qll

```diff
@@ -46,6 +46,26 @@ class FunctionInput extends TFunctionInput {
    */
   deprecated final predicate isInParameter(ParameterIndex index) { this.isParameter(index) }
 
+  /**
+   * Holds if this is the input value pointed to by a pointer parameter to a function, or the input
+   * value referred to by a reference parameter to a function, where the parameter has index
+   * `index`.
+   *
+   * Example:
+   * ```
+   * void func(int n, char* p, float& r);
+   * ```
+   * - `isParameterDeref(1, 1)` holds for the `FunctionInput` that represents the value of `*p` (with
+   *   type `char`) on entry to the function.
+   * - `isParameterDeref(2, 1)` holds for the `FunctionInput` that represents the value of `r` (with type
+   *   `float`) on entry to the function.
+   * - There is no `FunctionInput` for which `isParameterDeref(0, _)` holds, because `n` is neither a
+   *   pointer nor a reference.
+   */
+  predicate isParameterDeref(ParameterIndex index, int ind) {
```

Comment:
I don't understand what the `ind` parameter means here.

Reply:
Yeah, sorry. I've added a small amount of comments to it now. Here's the backstory: Currently on `main`, we cannot model a function like:
```cpp
int f(int** p) {
  return **p;
}
```
since we cannot say that there's flow from `**p` to the return value (we can only say that there's flow from `*p` by using `isParameterDeref(0)`). The use-use flow PR gives us the ability to model this correctly by adding another parameter to `isParameterDeref` that specifies the number of indirections. So we can say that there's flow from `*p` by using `isParameterDeref(0, 1)`, or flow from `**p` by using `isParameterDeref(0, 2)`.

The use-use flow PR requires these two argument predicates to be present (so that the code compiles), but all the code in non-experimental main will still just use the old predicates.

Does that make sense?

Reply:
Yes, that's fantastic, thanks for expanding on this!